File: app/src/main/java/com/walagran/wwf/ui/CreateGameActivity.java
package com.walagran.wwf.ui;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.widget.Button;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import com.walagran.wwf.R;
import com.walagran.wwf.ui.common.ControlsBar;
import com.walagran.wwf.ui.common.KeyboardEventListener;
import com.walagran.wwf.ui.common.KeyboardFragment;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Optional;

public class CreateGameActivity extends AppCompatActivity {
    KeyboardEventListener keyboardEventListener =
            new CreateGameKeyboardEventListener();
    ArrayList<TextView> letterViews = new ArrayList<>();
    ArrayList<Character> createdWord = new ArrayList<>(Arrays.asList('F', 'L'
            , 'A', 'S', 'H'));
    int cellInFocus = 0;
    int gameId = 0;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_create_game);

        setUpFragments();
        initializeLetterViews();
        initializeButtons();
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        Intent intent = new Intent(this, MainActivity.class);
        startActivity(intent);
        finish();
    }

    private void initializeButtons() {
        Button shareButton = findViewById(R.id.create_game_share_button);
        shareButton.setOnClickListener(view -> {
            Intent intent = new Intent();
            intent.setAction(Intent.ACTION_SEND);
            intent.setType("text/plain");
            getWord().ifPresent(word -> {
                intent.putExtra(Intent.EXTRA_SUBJECT, String.format("Sharing " +
                        "Game #%s", gameId));
                intent.putExtra(Intent.EXTRA_TEXT, word);
                startActivity(Intent.createChooser(intent, "Word"));
            });
        });
    }

    private Optional<String> getWord() {
        return (cellInFocus == 5) ? Optional.of(TextUtils.join("",
                createdWord)) : Optional.empty();
    }

    private void initializeLetterViews() {
        for (int i = 1; i < 6; i++) {
            letterViews
                    .add(findViewById(getResources()
                            .getIdentifier("create_game_created_letter_" + i,
                                    "id",
                                    getPackageName())));
        }
    }

    private void setUpFragments() {
        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.create_game_controls_bar,
                        ControlsBar.newInstance(String.format("Creating Game " +
                                "%s", gameId)))
                .commit();

        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.create_game_keyboard,
                        KeyboardFragment.newInstance(keyboardEventListener))
                .commit();
    }

    class CreateGameKeyboardEventListener implements KeyboardEventListener {
        @Override
        public void onAlphaKeyPressed(char alphabet) {
            if (cellInFocus == 5) {
                return;
            }
            letterViews.get(cellInFocus).setText(String.valueOf(alphabet));
            createdWord.set(cellInFocus, Character.toUpperCase(alphabet));
            cellInFocus++;
        }

        @Override
        public void onBackKeyPressed() {
            if (cellInFocus == 0) {
                return;
            }
            cellInFocus--;
            letterViews.get(cellInFocus).setText("");
        }

        @Override
        public void onEnterKeyPressed() {

        }
    }

}

File: app/src/main/java/com/walagran/wwf/ui/CreateGameActivity.kt
package com.walagran.wwf.ui

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.walagran.wwf.LetterState
import com.walagran.wwf.R
import com.walagran.wwf.ui.common.ControlsBar.Companion.newInstance
import com.walagran.wwf.ui.common.KeyboardEventListener
import com.walagran.wwf.ui.common.KeyboardFragment.Companion.newInstance
import java.util.*

class CreateGameActivity : AppCompatActivity() {
    private var keyboardEventListener: KeyboardEventListener =
        CreateGameKeyboardEventListener()

    private lateinit var shareButton: Button

    var letterViews = ArrayList<TextView>()
    var createdWord = ArrayList(listOf('F', 'L', 'A', 'S', 'H'))
    var cellInFocus = 0
    private var gameId = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_create_game)

        setUpBasicUIElements()

        initializeLetterViews()
    }

    override fun onBackPressed() {
        super.onBackPressed()
        val intent = Intent(this, MainActivity::class.java)
        startActivity(intent)
        finish()
    }

    private fun initializeLetterViews() {
        for (i in 1..5) {
            letterViews
                .add(findViewById(resources
                    .getIdentifier("create_game_created_letter_$i",
                        "id",
                        packageName)))
        }
    }

    private fun setUpBasicUIElements() {
        supportFragmentManager
            .beginTransaction()
            .replace(R.id.create_game_controls_bar,
                newInstance(String.format("Creating Game " +
                        "%s", gameId)))
            .commit()
        supportFragmentManager
            .beginTransaction()
            .replace(R.id.create_game_keyboard,
                newInstance(keyboardEventListener))
            .commit()

        shareButton = findViewById(R.id.create_game_share_button)
        shareButton.setOnClickListener {
            shareCreatedGame()
        }
    }

    private fun shareCreatedGame() {
        fun maybeGetInputWord(): Optional<String> {
            return if (cellInFocus == 5)
                Optional.of(TextUtils.join("", createdWord))
            else Optional.empty()
        }

        val intent = Intent()
        intent.action = Intent.ACTION_SEND
        intent.type = "text/plain"
        maybeGetInputWord().ifPresent {
            intent.putExtra(Intent.EXTRA_SUBJECT,
                String.format("Sharing Game #%s", gameId))
            intent.putExtra(Intent.EXTRA_TEXT, it)

            startActivity(Intent.createChooser(intent, "Word"))
        }
    }

    internal inner class CreateGameKeyboardEventListener :
        KeyboardEventListener {
        override fun onAlphaKeyPressed(alphabet: Char) {
            if (cellInFocus == 5) { return }

            letterViews[cellInFocus].text = alphabet.toString()
            createdWord[cellInFocus] = alphabet

            cellInFocus++
        }

        override fun onBackKeyPressed() {
            if (cellInFocus == 0) { return }

            cellInFocus--

            letterViews[cellInFocus].text = ""
        }

        override fun onEnterKeyPressed(): Map<Char, LetterState> {
            shareCreatedGame()
            return emptyMap()
        }
    }
}